File: OJ/leet523/leet523.py
# leet523

from typing import List


class Solution:
    def checkSubarraySum(self, nums: List[int], k: int) -> bool:
        sum = 0
        S = {}
        for i, x in enumerate(nums):
            sum += x 
            if k: sum %= k 
            if sum == 0 and i >= 1: return True
            if sum in S and i - S[sum] > 1: return True
            if sum not in S : 
                S[sum] = i
        return False


print(Solution().checkSubarraySum([23,2,4,6,7], k = 6))
print(Solution().checkSubarraySum([23,2,6,4,7], k = 6))
print(Solution().checkSubarraySum([23,2,0,4,7], k = 0))
print(Solution().checkSubarraySum([5,0,0], k = 0))

File: OJ/leet523/leet523.cpp
/*
leet523


*/


#include "../utils/LeetCpp.utils.hpp"
using namespace std;

class Solution {
public:
    bool checkSubarraySum(vector<int>& A, int k) {
        unordered_map<int, int> M;
        auto sum = 0;
        for (int i = 0; i < A.size(); i++)
        {
            sum += A[i];
            if (k) sum %= k;
            if (sum == 0 and i > 1) return true;
            if (M.find(sum) == M.end() ) M[sum] = i;
            else if (i - M[sum] > 1) return true;
        }
        return false;
    }
};


int main(int argc, char const *argv[])
{
    Solution sol;   
    
    return 0;
}
